src/NServiceBus.Core/Persistence/Msmq/SubscriptionStorage/MsmqSubscriptionStorage.cs

```diff
@@ -24,94 +23,82 @@ public void Dispose()
 
         public void Init()
         {
-            foreach (var m in storageQueue.GetAllMessages())
-            {
-                var messageTypeString = m.Body as string;
-                var messageType = new MessageType(messageTypeString); //this will parse both 2.6 and 3.0 type strings
-                var subscriber = Deserialize(m.Label);
+            var messages = storageQueue.GetAllMessages()
+                .OrderByDescending(m => m.ArrivedTime)
+                .ThenBy(x => x.Id) // ensure same order of messages with same timestamp accross all endpoints
+                .ToArray();
 
-                entries.Add(new Entry
+            lock (lookup)
+            {
+                foreach (var m in messages)
                 {
-                    MessageType = messageType,
-                    Subscriber = subscriber
-                });
-                AddToLookup(subscriber.TransportAddress, messageType, m.Id);
+                    var messageTypeString = m.Body as string;
+                    var messageType = new MessageType(messageTypeString); //this will parse both 2.6 and 3.0 type strings
+                    var subscriber = Deserialize(m.Label);
+
+                    Dictionary<MessageType, string> endpointSubscriptions;
+                    if (!lookup.TryGetValue(subscriber, out endpointSubscriptions))
+                    {
+                        lookup[subscriber] = endpointSubscriptions = new Dictionary<MessageType, string>();
+                    }
+
+                    if (endpointSubscriptions.ContainsKey(messageType))
+                    {
+                        // this message is stale and can be removed
+                        storageQueue.TryReceiveById(m.Id);
+                    }
+                    else
+                    {
+                        endpointSubscriptions[messageType] = m.Id;
+                    }
+                }
             }
         }
 
         public Task<IEnumerable<Subscriber>> GetSubscriberAddressesForMessage(IEnumerable<MessageType> messageTypes, ContextBag context)
         {
-            var messagelist = messageTypes.ToList();
-            lock (locker)
-            {
-                var result = (from e in entries
-                    from m in messagelist
-                    where e.MessageType == m
-                    select e)
-                    .Distinct(EntryComparer)
-                    .Select(e => e.Subscriber)
-                    .ToArray();
-
-                return Task.FromResult((IEnumerable<Subscriber>) result);
-            }
-        }
+            var messagelist = messageTypes.ToArray();
+            var result = new HashSet<Subscriber>();
 
-        public Task Subscribe(Subscriber subscriber, MessageType messageType, ContextBag context)
-        {
-            lock (locker)
+            lock (lookup)
```

Comment:
I wonder if the lock makes sense. Subscriptions are most often read and rarely updated that much. We could optimize for that by using a `ReaderWriterLockSlim` or apply reference swapping (need to double check if that really works).